crates/storage/provider/src/providers/snapshot/manager.rs

```diff
@@ -449,6 +449,91 @@ impl SnapshotProvider {
     pub fn directory(&self) -> &Path {
         &self.path
     }
+
+    /// Retrieves data from the database or snapshot, wherever it's available.
+    ///
+    /// # Arguments
+    /// * `segment` - The segment of the snapshot to check against.
+    /// * `index_key` - Requested index key, usually a block or transaction number.
+    /// * `fetch_from_snapshot` - A closure that defines how to fetch the data from the snapshot
+    ///   provider.
+    /// * `fetch_from_database` - A closure that defines how to fetch the data from the database
+    ///   when the snapshot doesn't contain the required data or is not available.
+    pub fn get_with_snapshot_or_database<T, FS, FD>(
+        &self,
+        segment: SnapshotSegment,
+        number: u64,
+        fetch_from_snapshot: FS,
+        fetch_from_database: FD,
+    ) -> ProviderResult<Option<T>>
+    where
+        FS: Fn(&SnapshotProvider) -> ProviderResult<Option<T>>,
+        FD: Fn() -> ProviderResult<Option<T>>,
+    {
+        // If there is, check the maximum block or transaction number of the segment.
+        let snapshot_upper_bound = match segment {
+            SnapshotSegment::Headers => self.get_highest_snapshot_block(segment),
+            SnapshotSegment::Transactions | SnapshotSegment::Receipts => {
+                self.get_highest_snapshot_tx(segment)
+            }
+        };
+
+        if snapshot_upper_bound.map_or(false, |snapshot_upper_bound| snapshot_upper_bound >= number)
+        {
+            return fetch_from_snapshot(self)
+        }
+        fetch_from_database()
+    }
+
+    /// Gets data within a specified range, potentially spanning different snapshots and database.
+    ///
+    /// # Arguments
+    /// * `segment` - The segment of the snapshot to query.
+    /// * `block_range` - The range of data to fetch.
+    /// * `fetch_from_snapshot` - A function to fetch data from the snapshot.
+    /// * `fetch_from_database` - A function to fetch data from the database.
+    /// * `predicate` - A function used to evaluate each item in the fetched data. Fetching is
+    ///   terminated when this function returns false, thereby filtering the data based on the
+    ///   provided condition.
+    pub fn get_range_with_snapshot_or_database<T, P, FS, FD>(
```

Comment:
this a duplicate from another function which was part of `DatabaseProvider`, will dedup on a follow-up